src/dataflow-types/src/client.rs

```diff
@@ -834,28 +874,44 @@ pub trait Reconnect {
     async fn reconnect(&mut self);
 }
 
+/// Trait for clients that can connect to an address
+pub trait FromAddr {
+    fn from_addr(addr: String) -> Self;
+}
+
 /// A convenience type for compatibility.
 #[derive(Debug)]
-pub struct RemoteClient<C, R>
+pub struct RemoteClient<C, R, G = tcp::TcpClient<C, R>>
 where
     (C, R): partitioned::Partitionable<C, R>,
     C: fmt::Debug + Send,
     R: fmt::Debug + Send,
+    G: GenericClient<C, R> + Reconnect + FromAddr,
 {
-    client: partitioned::Partitioned<tcp::TcpClient<C, R>, C, R>,
+    client: partitioned::Partitioned<G, C, R>,
 }
 
-impl<C, R> RemoteClient<C, R>
+pub type ComputedRemoteClient<T> =
+    RemoteClient<ComputeCommand<T>, ComputeResponse<T>, tcp::GrpcComputedClient>;
+
+pub type StoragedRemoteClient<T> = RemoteClient<
+    StorageCommand<T>,
+    StorageResponse<T>,
+    tcp::TcpClient<StorageCommand<T>, StorageResponse<T>>,
+>;
+
+impl<C, R, G> RemoteClient<C, R, G>
 where
     (C, R): partitioned::Partitionable<C, R>,
     C: fmt::Debug + Send,
     R: fmt::Debug + Send,
+    G: GenericClient<C, R> + Reconnect + FromAddr,
 {
     /// Construct a client backed by multiple tcp connections
     pub fn new(addrs: &[impl tokio::net::ToSocketAddrs + std::fmt::Display]) -> Self {
         let mut remotes = Vec::with_capacity(addrs.len());
         for addr in addrs.iter() {
-            remotes.push(tcp::TcpClient::new(addr.to_string()));
+            remotes.push(G::from_addr(addr.to_string()));
```

Comment:
Change `from_addr` to take a `tokio::net::ToSocketAddrs` instead of round-tripping through a string -- if this is possible with the gRPC library.

Reply:
For the tonic connect, I need something that can be converted into and Endpoint, which in turn contains a `Uri` (  https://docs.rs/tonic/latest/src/tonic/transport/channel/endpoint.rs.html#25-44 ). So for gRPC, the to string roundtrip seems unavoidable. But I can do the conversion in the `impl FromAddr for GrpcComputedClient` and not outside.

Reply:
It's probably not worth it, TcpClient could be changed to keep ToSockAddrs around (it also stores a string right now), but we hopefully replace all of them with gRPC based ones in the near future anyway.